include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h

```diff
@@ -0,0 +1,661 @@
+    }
+
+
+
+    // Multiplications
+
+    template <typename Number, typename Node>
+    void
+    SparseMatrix<Number, Node>::vmult(Vector<Number>       &dst,
+                                      const Vector<Number> &src) const
+    {
+      Assert(&src != &dst, ExcSourceEqualsDestination());
+      Assert(matrix->isFillComplete(), ExcMatrixNotCompressed());
+      Assert(src.trilinos_rcp()->getMap()->isSameAs(*matrix->getDomainMap()),
+             ExcColMapMissmatch());
+      Assert(dst.trilinos_rcp()->getMap()->isSameAs(*matrix->getRangeMap()),
+             ExcDomainMapMissmatch());
+      matrix->apply(*src.trilinos_rcp(), *dst.trilinos_rcp());
+    }
+
+
+
+    template <typename Number, typename Node>
+    void
+    SparseMatrix<Number, Node>::Tvmult(Vector<Number>       &dst,
+                                       const Vector<Number> &src) const
+    {
+      Assert(&src != &dst, ExcSourceEqualsDestination());
+      Assert(matrix->isFillComplete(), ExcMatrixNotCompressed());
+      Assert(dst.trilinos_rcp()->getMap()->isSameAs(*matrix->getDomainMap()),
+             ExcColMapMissmatch());
+      Assert(src.trilinos_rcp()->getMap()->isSameAs(*matrix->getRangeMap()),
+             ExcDomainMapMissmatch());
+      matrix->apply(*src.trilinos_rcp(), *dst.trilinos_rcp(), Teuchos::TRANS);
+    }
+
+
+
+    template <typename Number, typename Node>
+    void
+    SparseMatrix<Number, Node>::vmult_add(Vector<Number>       &dst,
+                                          const Vector<Number> &src) const
+    {
+      Assert(&src != &dst, ExcSourceEqualsDestination());
+      Assert(matrix->isFillComplete(), ExcMatrixNotCompressed());
+      Assert(src.trilinos_rcp()->getMap()->isSameAs(*matrix->getDomainMap()),
+             ExcColMapMissmatch());
+      Assert(dst.trilinos_rcp()->getMap()->isSameAs(*matrix->getRangeMap()),
+             ExcDomainMapMissmatch());
+      matrix->apply(*src.trilinos_rcp(),
+                    *dst.trilinos_rcp(),
+                    Teuchos::NO_TRANS,
+                    Teuchos::ScalarTraits<Number>::one(),
+                    Teuchos::ScalarTraits<Number>::one());
+    }
+
+
+
+    template <typename Number, typename Node>
+    void
+    SparseMatrix<Number, Node>::Tvmult_add(Vector<Number>       &dst,
+                                           const Vector<Number> &src) const
+    {
+      Assert(&src != &dst, ExcSourceEqualsDestination());
+      Assert(matrix->isFillComplete(), ExcMatrixNotCompressed());
+      Assert(dst.trilinos_rcp()->getMap()->isSameAs(*matrix->getDomainMap()),
+             ExcColMapMissmatch());
+      Assert(src.trilinos_rcp()->getMap()->isSameAs(*matrix->getRangeMap()),
+             ExcDomainMapMissmatch());
+      matrix->apply(*src.trilinos_rcp(),
+                    *dst.trilinos_rcp(),
+                    Teuchos::NO_TRANS,
+                    Teuchos::ScalarTraits<Number>::one(),
+                    Teuchos::ScalarTraits<Number>::one());
```

Comment:
I think you want the third argument here to be `TRANS`, right?

Reply:
Thank you for spotting this one!

This one shows again that the next step is to add tests to cover all these functions.

Reply:
Yes. I gave some thought about how we might approach this and will give it a try once we're done merging these two PRs for matrices and vectors.